torch/_masked/__init__.py

```diff
@@ -390,21 +396,71 @@ def _canonical_dim(dim: DimOrDims, ndim: int) -> Tuple[int, ...]:
 
 def _input_mask(input: Tensor, *args, **kwargs) -> Tensor:
     """Return canonical input mask.
-    Canonical input mask is a boolean tensor with the same shape as
-    input and with (broadcasted) content of mask, if specified.
+
+    For strided inputs, the canonical input mask is a boolean tensor
+    with the same shape as input and with (broadcasted) content of
+    mask, if specified.
+
+    For sparse COO inputs, the canonical input mask is a sparse
+    coalesced tensor with the same shape and indices as the coalesced
+    :attr:`input` tensor.  The values of the canonical input masks are
+    all True, unless the :attr:`mask` tensor dtype is torch.bool, in
```

Comment:
Shouldn't the presence of a mask always influence the behavior even if it has other dtypes (or perhaps other dtypes aren't allowed)?

Reply:
When sparse mask is specified, its indices define the mask and, in general, the values are intentionally ignored (because one may specify `input` as `mask`). So, there is no reason to disallow other dtypes because dtype is related to values and hence also ignored because we re-map mask [when dtype is not bool]:
```python
values = mask.values()
mask = torch.sparse_coo_tensor(
    mask.indices(),
    values.new_ones(values.shape, dtype=torch.bool),
    mask.shape)._coalesced_(True)
```

The exception is "dtype is bool" case where the values of mask have significance because we use [in the case of masked sum]:
```
mask_input = input * mask
```
where non-true mask values mask-out the corresponding input values that otherwise would be masked-in by mask indices.

If I now think about masked product, for instance, on a sparse tensor, then taking into account the boolean values of a mask requires implementing `torch.where` support on sparse tensors such that its result would have the same sparsity pattern as input. This should be addressed in a follow-up PR.

@cpuhrsch does this answer your question?

Reply:
The only thing that bothers me here is that the mask values might have been carefully chosen to have some `False` entries. The user could have cause some materialized zeros via some prior op (like relu) and now wants to apply a reduction that ignores materialized zeros as well. I think we can address that via coalesce. It's also ok to require the indices of the mask to match the indices of the input for now and otherwise throw. Then the user will just have to do it themselves and we can add support later on.

I also think we should disallow masks without dtype bool for now, because we've run into issues in the past in MHA with float masks etc. Bool is less ambiguous.

As a side note, it should be very useful eventually to support 0-dim values. That'd allow us to say "all values are True" without having to store them explicitly. This is an optimization for later on though I think. So then something like ```ones_like(sparse_input)``` could just reference the indices of `sparse_input`.